### PR TITLE
Merge 2.11 to 2.12 [ci: last-only]

### PR DESCRIPTION
**Pull request by lrytz**
```
$ git log --decorate --oneline -1 2.11.x | cat
aaf7bc0 (upstream/2.11.x, 2.11.x) Merge pull request #5190 from szeiger/wip/validate-test-use-sbt

$ git log --decorate --oneline -1 2.12.x | cat
79e7334 (HEAD -> 2.12.x, upstream/2.12.x, origin/2.12.x) Merge pull request #5202 from lrytz/lineNumbers

$ export MB=$(git merge-base 2.12.x 2.11.x)

$ git checkout 2.12.x
$ git checkout -b merge-2.11-to-2.12-june-19

$ git log --graph --oneline --decorate $MB...2.11.x | cat
*   aaf7bc0 (upstream/2.11.x, 2.11.x) Merge pull request #5190 from szeiger/wip/validate-test-use-sbt
|\
| * c2c08a4 Remove bnd warnings from log output
| * f2d0f1e Use sbt for PR validation
|/
*   91b6944 Merge pull request #5219 from som-snytt/issue/9245
|\
| * 3f3a3bb SI-9245 Fresher name in Try and test
*   ab06ce8 Merge pull request #5213 from som-snytt/issue/stringof-pariterable
|\
| * be9c0fe SI-9737 [no-merge] Backport stringOf ParIterable
* f775e1f Merge pull request #5223 from FelAl/typo_fix
* c7e2b28 Typo fix

$ git merge f775e1f
CONFLICT (content): Merge conflict in build.sbt
  >> small conflict, easy fix

$ git merge ab06ce8 -s ours
Merge made by the 'ours' strategy.

$ git merge 91b6944

$ git merge aaf7bc0
CONFLICT (content): Merge conflict in versions.properties
  >> partest.version.number

CONFLICT (rename/delete): test/disabled/run/t7933.scala deleted in HEAD and renamed in aaf7bc0.
CONFLICT (rename/delete): test/disabled/run/t7933.check deleted in HEAD and renamed in aaf7bc0.
CONFLICT (rename/delete): test/disabled/run/t7843-jsr223-service.scala deleted in HEAD and renamed in aaf7bc0.
CONFLICT (rename/delete): test/disabled/run/t7843-jsr223-service.check deleted in HEAD and renamed in aaf7bc0.
  >> removed (those tests were removed in 2.12.x in 3cddeaa)

CONFLICT (content): Merge conflict in scripts/jobs/validate/test
  >> nothing complicated (2.12 passed -opt:l:classpath instead of -optimise)

CONFLICT (content): Merge conflict in build.sbt
  >> some conflicts in dependencies (2.12 doesn't have akka actors, continuations, but JOL)
  >> a conflict because 2.12 doesn't build forkjoin
  >> a conflict because 2.12 hase a new command to generated product/function/.. sources (bfb7a74)
```

### CODE REVIEW

**Comment by lrytz** [comment]
review by @szeiger

**Comment by lrytz** [comment]
```
[error] bad option: '-opt:l:classpath'
```

the issue is that we're passing this option to the starr compiler (2.12.0-M4) which doesn't understand it yet.. in principle it should only be passed while building the second stage.

i pushed another commit that uses the deprecated `-optimize` instead, which emits a warning and enables `-opt:l:classpath`. however, the warning will probably fail some partest tests / `.check` files.

File: scripts/jobs/validate/test [hunk 1]
```diff
@@ -1,17 +1,36 @@
...
+    # and run JUnit tests, partest, OSGi tests, MiMa and scaladoc
+    $SBT_CMD \
+       -Dstarr.version=$scalaVersion \
+       --warn \
+       "setupValidateTest $prRepoUrl" \
+       $testExtraArgs \
+       "test" \
+       "partest run pos neg jvm" \
+       "partest res scalap specialized scalacheck" \
+       "partest instrumented presentation" \
```
Comment (szeiger) [inline scripts/jobs/validate/test:26]: I've only seen PermGen errors on 2.11 when running too many partest tests in one go, but the heap error here looks similar. Splitting this up into two runs may work better:

```
      "partest instrumented" \
      "partest presentation" \
```

**Comment by lrytz** [comment]
hmm, no, that didn't help..

**Comment by lrytz** [comment]
I can reproduce the OOME locally, when running the `presentation` tests with `-opt:l:classpath` (the stack trace shows that the error happens in the optimizer).

I'm not sure that partest correctly picks up the java options defined here: https://github.com/scala/scala/pull/5237/commits/f2d0f1e85d8b348cd2506a45502e4e59f9ec8e49#diff-fdc3abdfd754eeb24090dbd90aeec2ceR691. They do show up in the log output:

```
Java options are:    -Xmx1024M -Xms64M -XX:MaxPermSize=128M
```

but when I change them to something larger (`-Xmx4096M`), the process that crashes with OOME does not seem to allocate more memory than before.

**Comment by lrytz** [comment]
to reproduce, run

```
sbt -Dstarr.version="2.12.0-6be982e-SNAPSHOT" 'setupValidateTest https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots/' 'partest presentation'
```

note that `setupValidateTest` also enables the optimizer.

**Comment by retronym** [comment]
@lrytz perhaps try passing invalid JVM options to make sure they are passed down?

**Comment by lrytz** [comment]
Good idea, I tried that. However the result is the same, we still get OOMEs while compiling the first tests. After that error, the other tests are compiled an run, and they do fail because of the invalid flag.

So it seems that the jvm flags are only passed when running the test, but not to the JVM that compiles the tests.

**Comment by lrytz** [comment]
just saw that there's a separate setting for the java options of the JVM that's running partest, changing this helped locally. added a commit 62035e9.

**Comment by szeiger** [comment]
If it's the actual `scalac` command that doesn't get the options, I don't think this will help. I'm still trying to figure out how they are supposed to be passed down in partest. Maybe there's a problem with `DirectCompiler` similar to the one we had for `scalacOpts`.

**Comment by lrytz** [comment]
IRC, partest runs the compiler (actually multiple instances of it) in its own JVM

**Comment by szeiger** [comment]
Ah, I think you're right. It still uses `scalacOpts` because it creates a new `Global` with these settings but it shouldn't need separate `javaOpts`. We should know for sure in a few minutes :)

**Comment by lrytz** [comment]
it helped, all partest tests passed. now we just need to fix osgi :)

```
[error] Test tools.test.osgi.libonly.BasicLibraryTest.everythingLoads failed: java.lang.ClassNotFoundException: scala.collection.GenTraversableOnce, took 0.537 sec
```

**Comment by szeiger** [comment]
I haven't cross-checked with 2.11.x yet but this looks suspicious:

```
*** (o.o.p.e.n.i.NativeTestContainer) Bundle [org.scala-lang.scala-library_2.12.0.v20160621-120706-62035e9 [14]] is not resolved
```

Maybe we broke something in the bundle creation with an earlier merge of the OSGi functionality into 2.12.x?

**Comment by szeiger** [comment]
So far unable to reproduce this locally. I had some other errors (indicating duplicate bundles) because I didn't run `osgiTestFelix/clean` before `osgiTestFelix/test`. Now that I did, the tests passed.

**Comment by lrytz** [comment]
same here, after `osgiTestFelix/clean` it passed locally, but not before. i don't have any intuition what could cause this (on the ci). should we just add the clean command to the script?

**Comment by szeiger** [comment]
It shouldn't be needed on CI. There shouldn't be anything in the folder that we don't want. The error message when this occured locally was also pretty much the opposite from what we've seen on CI (multiple instances of a bundle instead of `ClassNotFoundException`)

**Comment by szeiger** [comment]
/rebuild

Partest failures in:

```
# Failed test paths (this command will update checkfiles)
partest --update-check \
  /home/jenkins/workspace/scala-2.12.x-validate-test/test/files/run/annotatedRetyping.scala \
  /home/jenkins/workspace/scala-2.12.x-validate-test/test/files/run/class-symbol-contravariant.scala \
  /home/jenkins/workspace/scala-2.12.x-validate-test/test/files/run/compiler-asSeenFrom.scala \
  /home/jenkins/workspace/scala-2.12.x-validate-test/test/files/run/constant-type.scala \
  /home/jenkins/workspace/scala-2.12.x-validate-test/test/files/run/constrained-types.scala \
  /home/jenkins/workspace/scala-2.12.x-validate-test/test/files/run/ctries-new \
  /home/jenkins/workspace/scala-2.12.x-validate-test/test/files/run/ctries-old \
  /home/jenkins/workspace/scala-2.12.x-validate-test/test/files/run/existential-rangepos.scala \
  /home/jenkins/workspace/scala-2.12.x-validate-test/test/files/run/existentials-in-compiler.scala \
  /home/jenkins/workspace/scala-2.12.x-validate-test/test/files/run/existentials.scala
```

Can't reproduce locally. Spurious errors?

**Comment by lrytz** [comment]
it would be great if the jenkins output would show the partest log output of failed tests (see https://scala-ci.typesafe.com/job/scala-2.12.x-validate-test/2389/consoleFull), it used to be the case with ant. not sure what flags we need to pass to partest for that..

**Comment by som-snytt** [comment]
@lrytz `--show-diff` is the diff, but somewhere I recently restored `--show-log`, which will show log of crashed tests, where you want the stack trace.

**Comment by szeiger** [comment]
Okay, now we got a totally different set of failing tests...

**Comment by szeiger** [comment]
@lrytz I'm trying to figure out why ant shows more output. I'll push a commit to enable it in sbt, too. Maybe this will give us an idea of what the failures are.

**Comment by szeiger** [comment]
Now that I've added the failure logging in #5242, partest succeeds (as expected). But I was finally able to reproduce the OSGi failure locally. I'm trying to narrow it down.

**Comment by szeiger** [comment]
I was able to reproduce the failure on every test run locally. Upgrading to newer versions of the OSGi test frameworks (same as in the ant build on 2.12.x) seems to fix it. Let's see what Jenkins has to say...

**Comment by szeiger** [comment]
OSGi tests working again! Now disabling MiMa and hoping for a green build (or some of those spurious errors, now with more debug output)

**Comment by lrytz** [comment]
cool! closing this one in favor of #5242

**Pull request closed without merging**